lib/lua/mark_eats/musicutil.lua

```diff
@@ -0,0 +1,197 @@
+        end
+        if found then break end
+      end
+    end
+  end
+  
+  local scale_data =  MusicUtil.SCALES[scale_type]
+  if not scale_data then return nil end
+
+  -- Generate output array
+  local output = {}
+  local scale_len = #scale_data.intervals
+  local note_num
+  for i = 0, octaves * scale_len - 1 do
+    if i > 0 and i % scale_len == 0 then
+      root_num = root_num + scale_data.intervals[scale_len]
+    else
+      note_num = root_num + scale_data.intervals[i % scale_len + 1]
+      if note_num > 127 then break
+      else table.insert(output, note_num) end
+    end
+  end
+  return output
+end
+
+--- Snap MIDI note numbers to an array of note numbers.
+-- @param note_nums MIDI note number (0-127) or array of note numbers.
+-- @param snap_array Array of note numbers to snap to, must be in low to high order.
+-- @return Array of adjusted note numbers or a single note number.
+function MusicUtil.snap_notes_to_array(note_nums, snap_array)
+  if not note_nums or not snap_array then return nil end
+  if type(note_nums) == "number" then note_nums = {note_nums} end
+  
+  local delta
+  local prev_delta
+  local note_nums_len = #note_nums
+  local snap_array_len = #snap_array
+  for n = 1, note_nums_len do
+    if snap_array_len == 1 then
+      note_nums[n] = snap_array[1]
+    elseif note_nums[n] >= snap_array[snap_array_len] then
+      note_nums[n] = snap_array[snap_array_len]
+    else
+      prev_delta = math.huge
+      for s = 1, snap_array_len + 1 do
+        if s > snap_array_len then
+          note_nums[n] = note_nums[n] + prev_delta
+          break
+        end
+        delta = snap_array[s] - note_nums[n]
+        if delta == 0 then
+          break
+        elseif math.abs(delta) >= math.abs(prev_delta) then
+          note_nums[n] = note_nums[n] + prev_delta
+          break
+        end
+        prev_delta = delta
+      end
+    end
+  end
+  if note_nums_len == 1 then return note_nums[1]
+  else return note_nums end
+end
+
+--- Convert MIDI note numbers to names.
+-- @param note_nums MIDI note number (0-127) or array of note numbers.
+-- @param[opt] include_octave Include octave number in return string if set to true.
+-- @return Name string (eg, "C#3") or array of strings.
+function MusicUtil.note_nums_to_names(note_nums, include_octave)
+  if not note_nums then return nil end
+  if type(note_nums) == "number" then note_nums = {note_nums} end
+  local note_nums_len = #note_nums
+  local output = {}
+  for i = 1, note_nums_len do
+    local name = MusicUtil.NOTE_NAMES[note_nums[i] % 12 + 1]
+    if include_octave then name = name .. math.floor(note_nums[i] / 12 - 1) end
+    output[i] = name
+  end
+  if note_nums_len == 1 then return output[1]
+  else return output end
+end
+
+--- Convert MIDI note numbers to frequencies.
+-- @param note_nums MIDI note number (0-127) or array of note numbers.
+-- @return Frequency number in Hz or array of frequencies.
+function MusicUtil.note_nums_to_freqs(note_nums)
```

Comment:
Was wondering if there's any particular reason this takes multiple notes as input? Can't the script just call the function as many times as it needs it?
And if there is merit to having it as a library shouldn't it just be a separate function that calls the simple/one note input function multiple times?

Reply:
It was just for convenience, but you're right, it's unusual to have methods returning varying types like that, I got a little over enthusiastic with the dynamic typing there.  😄 
I just pushed a commit that refactors these into 'single item' methods and 'array methods' for clarity. Thx for the feedback!